core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -291,6 +289,132 @@ object Uri {
+      hash
+    }
+
+    def render(writer: Writer): writer.type = {
+      val start = if (absolute) "/" else ""
+      writer << start << segments.iterator.mkString("/")
+      if (endsWithSlash) writer << "/" else writer
+    }
+
+    override val renderString: String = super.renderString
+    override def toString: String = renderString
+
+    def /(segment: Path.Segment): Path = addSegment(segment)
+    def addSegment(segment: Path.Segment): Path =
+      addSegments(List(segment))
+    def addSegments(value: List[Path.Segment]): Path =
+      Path(this.segments ++ value, absolute = absolute || this.segments.isEmpty)
+
+    def normalize: Path = Path(segments.filterNot(_.isEmpty))
+
+    /* Merge paths per RFC 3986 5.2.3 */
+    def merge(path: Path): Path = {
+      val merge = segments.init
+      Path(merge ++ path.segments, absolute = absolute, endsWithSlash = path.endsWithSlash)
+    }
+
+    def startsWith(path: Path): Boolean = segments.startsWith(path.segments)
+
+    def startsWithString(path: String): Boolean = startsWith(Path.fromString(path))
+
+    def indexOf(path: Path): Option[Int] =
+      Some(segments.indexOfSlice(path.segments)).filterNot(_ == -1)
+
+    def indexOfString(path: String): Option[Int] = indexOf(Path.fromString(path))
+
+    def splitAt(idx: Int): (Path, Path) = {
+      val (start, end) = segments.splitAt(idx)
+      Path(start, absolute = absolute) -> Path(end, endsWithSlash = endsWithSlash)
+    }
+  }
+
+  object Path {
+    val empty = Path(Nil)
+    val Root = Path(Nil, absolute = true)
+
+    final class Segment private (val encoded: String) {
+      def isEmpty = encoded.isEmpty
+
+      override def equals(obj: Any): Boolean =
+        obj match {
+          case s: Segment => s.encoded == encoded
+        }
+
+      override def hashCode(): Int = encoded.hashCode
+
+      def decoded: String = Uri.decode(encoded)
+
+      override val toString: String = encoded
+    }
+
+    object Segment extends (String => Segment) {
+      def apply(value: String): Segment = new Segment(pathEncode(value))
+      def encoded(value: String): Segment = new Segment(value)
+    }
+
+    /**
+      * This constructor allows you to construct the path directly.
+      * Each path segment needs to be encoded for it to be used here.
+      *
+      * @param segments the segments that this path consists of. MUST be Urlencoded.
+      * @param absolute if the path is absolute. I.E starts with a "/"
+      * @param endsWithSlash if the path is a "directory", ends with a "/"
+      * @return a Uri.Path that can be used in Uri, or by itself.
+      */
+    def apply(
+        segments: List[Segment],
+        absolute: Boolean = false,
+        endsWithSlash: Boolean = false): Path =
+      new Path(segments, absolute, endsWithSlash)
+
+    def unapply(path: Path): Some[(List[Segment], Boolean, Boolean)] =
+      Some((path.segments, path.absolute, path.endsWithSlash))
+
+    def fromString(fromPath: String): Path =
```

Comment:
rename this to apply, then let `object Path extends (String => Path)` ?
Ref @rossabaker comment above.

Reply:
My comment was more general, for all the times we avoid a case class so we can validate the argument.

I don't really like overloading `apply`.  I think I'd probably leave it as it is in this case.